### ISSUE TITLE
ub_overlap_ag_fprop disabled during activation recomputation due to is_grad_enabled check

### ISSUE THREAD

**Issue by SophusDavid** [NVIDIA/TransformerEngine#2902]
# Describe the bug
When activation recomputation is enabled, the forward pass is re-executed under torch.no_grad(). This causes is_grad_enabled to be False, which unconditionally disables the Userbuffers all-gather overlap (ub_overlap_ag_fprop) during the recompute forward pass.Relevant code at [layernorm_linear.py#L195](https://github.com/NVIDIA/TransformerEngine/blob/262bc6cfe1bb8d20b9367c1e5339af78e7090b19/transformer_engine/pytorch/module/layernorm_linear.py#L195):

```
ub_overlap_ag_fprop = (
    ub_overlap_ag_fprop and is_grad_enabled and not return_layernorm_output
)
```

The intent of checking is_grad_enabled here appears to be: "only overlap communication in forward if we will later need the backward pass." However, during activation recomputation, the forward pass is part of backward (it reconstructs activations needed for gradient computation). The recompute forward still performs the same GEMM + all-gather operations, so disabling comm overlap here causes a significant performance regression — the all-gather and GEMM become serialized instead of overlapped.This means enabling activation recomputation (e.g., via Megatron-LM's --recompute-activations) silently degrades TP communication overlap performance, which is counterproductive since recomputation is typically used in memory-constrained scenarios where maximizing overlap is critical.Steps/Code to reproduce bug
```

import torch
import transformer_engine.pytorch as te

# Minimal setup with TP overlap enabled
# (requires multi-GPU with userbuffers configured)

model = te.LayerNormLinear(
    hidden_size, ffn_hidden_size,
    params_dtype=torch.bfloat16,
    # ... with ub_overlap_ag enabled via env/config
)

# Case 1: Normal forward — ub_overlap_ag_fprop is active
with torch.enable_grad():
    out = model(inp)  # comm overlap works ✓

# Case 2: Recompute forward — ub_overlap_ag_fprop is disabled
with torch.no_grad():
    out = model(inp)  # comm overlap disabled ✗ (is_grad_enabled=False)
```

In practice, this is triggered by Megatron-LM's activation recomputation via CheckpointFunction, which calls the forward pass inside torch.no_grad().Expected behaviorub_overlap_ag_fprop should remain active during the recompute forward pass. The comm+GEMM overlap is a pure performance optimization for the forward computation itself — it does not depend on whether gradients are being tracked.A possible fix:
```
# Option 1: Remove the is_grad_enabled check entirely for fprop overlap
ub_overlap_ag_fprop = (
    ub_overlap_ag_fprop and not return_layernorm_output
)

# Option 2: Also allow overlap during recompute phase
ub_overlap_ag_fprop = (
    ub_overlap_ag_fprop
    and (is_grad_enabled or in_fp8_activation_recompute_phase())
    and not return_layernorm_output
)
```

**Comment by Saad-Mallebhari** [comment]
Hi @SophusDavid

Thanks for the detailed issue report and the exact reproduction snippet.

Root Cause

In `transformer_engine/pytorch/module/layernorm_linear.py` (line 195, as well as `layernorm_mlp.py` and `linear.py`), `ub_overlap_ag_fprop` explicitly checks `torch.is_grad_enabled()`. When activation recomputation (gradient checkpointing) is enabled in frameworks like Megatron-LM, the forward pass is re-executed inside a `torch.no_grad()` context to reconstruct intermediate activations for backward gradient calculation. Because `is_grad_enabled` evaluates to `False`, Userbuffers All-Gather communication overlap is unconditionally disabled during recomputation, serializing communication and GEMM computation.

Proposed Solution

1. Refactor `ub_overlap_ag_fprop` evaluation in `transformer_engine/pytorch/module/layernorm_linear.py` (and related module wrappers `layernorm_mlp.py` and `linear.py`) to remove the strict `is_grad_enabled` requirement for forward communication overlap.
2. Ensure Userbuffers All-Gather overlap remains active during both standard forward and recompute forward passes regardless of `torch.is_grad_enabled()` status.
3. Add regression tests in `tests/pytorch/test_layernorm_linear.py` verifying that `ub_overlap_ag_fprop` evaluates to `True` under `torch.no_grad()` contexts when userbuffers overlap is enabled in configuration.

Would you like me to open a Pull Request with this overlap fix and accompanying regression tests?

**Comment by ravimajeti** [comment]
Hey @Saad-Mallebhari 

Thanks for looking into this and proposing a fix. I agree that removing the `is_grad_enabled` restriction may be the right direction.

Before opening a PR, I think it would be useful to validate the change against the actual UserBuffers communication path rather than only testing that `ub_overlap_ag_fprop` remains enabled under `torch.no_grad()`.

In particular, I'd like to verify that:

* the UserBuffers All-Gather path is actually exercised when gradients are disabled;
* outputs remain numerically correct compared with the existing non-overlap path;
* checkpointed training/recomputation continues to produce correct outputs and gradients; and
* there aren't any UserBuffers assumptions around buffer lifetime, synchronization, or backward execution that motivated the original `is_grad_enabled` guard.

I'm currently working on reproducing the issue on a multi-GPU setup and testing the proposed change. Once I have the reproduction and correctness results, I can share them here and, if everything looks good, put together the fix and regression tests.

**Comment by Saad-Mallebhari** [comment]
Hi @ravimajeti

Thanks for following up and taking on the multi-GPU verification pass.

Validating those four dimensions against the native UserBuffers transport layer makes total sense. Because activation recomputation re-executes module forward passes inside `torch.no_grad()`, removing `is_grad_enabled` must ensure that:

1. UserBuffers All-Gather handles and communicator ring buffers remain valid without early deallocation during `no_grad` forward execution.
2. Intermediates reconstructed for backward gradient computation maintain strict numerical equivalence against non-overlapped baseline passes.
3. Stream synchronization hooks between UserBuffers C++ primitives and PyTorch's autograd engine avoid memory aliasing or race conditions during recomputation steps.

Looking forward to your multi-GPU benchmark and correctness findings on this.
